newrelic_plugin_agent/plugins/couchbase.py

```diff
@@ -0,0 +1,158 @@
+            for metric in [m for m in self.METRICS if m['type'] == typ]:
+                if typ == 'cluster':
+                    items = stats
+                    name_key = 'name'
+                elif typ == 'nodes':
+                    items = stats['nodes']
+                    name_key = 'hostname'
+                elif typ == 'buckets':
+                    items = stats
+                    name_key = 'name'
+
+                # count score for scoreboard metrics,
+                # otherwise just add the gauge value from the API repsonse
+                if metric.get('scoreboard', False):
+                    self.add_gauge_value(
+                        '%s/_scoreboard/%s' % (typ, metric['label']),
+                        metric['suffix'],
+                        self._get_metric_score(metric, items))
+                else:
+                    # add gauge value for current metric.
+                    # NOTE cluster metrics are not repeated,
+                    # as there is a single cluster.
+                    # nodes and bucket metrics are repeated,
+                    # as there are (usually) several of them
+                    if typ == 'cluster':
+                        self._add_gauge_value(
+                            metric, typ, items[name_key], items)
+                    else:
+                        for item in items:
+                            self._add_gauge_value(
+                                metric, typ, item[name_key], item)
+
+    def _add_gauge_value(self, metric, typ, name, items):
+        """Adds a gauge value for a nested metric.
+
+        Some stats are missing from memcached bucket types,
+        thus we use dict.get()
+
+        :param dict m: metric as defined at the top of this class.
+        :param str typ: metric type: cluster, nodes, buckets.
+        :param str name: cluster, node, or bucket name.
+        :param dict items: stats to lookup the metric in.
+
+        """
+        label = metric['label']
+        value = items
+        for key in label.split('.'):
+            value = value.get(key, 0)
+        self.add_gauge_value('%s/%s/%s' % (typ, name, label),
+                             metric['suffix'], value)
+
+    def _get_metric_score(self, metric, items):
+        """Calculate metric score for scoreboard type metrics.
+
+        A "scoreboard" metric is a counter of a specific value for a field.
+        E.g How many times the pair {"status": "healthy"} is found in a
+        list of object.
+
+        NOTE that scoreboard metrics are meant to be used on lists of objects,
+        i.e. a list of node objects.
+
+        :param dict metric: metric to calculate score for.
+        :param items metric: stats to calculate score from.
+
+        :rtype: int
+
+        """
+        score = 0
+        label = metric['label']
+        score_value = metric['score_value']
+        for d in items:
+            score += 1 if d[label] == score_value else 0
```

Comment:
nice!
does the second option cache the variables, or fetches them from the dict every time?